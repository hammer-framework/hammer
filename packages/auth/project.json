{
  "namedInputs": {
    "prod": [
      "!{projectRoot}/**/*.test.{js,jsx,ts,tsx}",
      "{workspaceRoot}/babel.config.js",
      "{workspaceRoot}/tsconfig.build.json",
      {
        "runtime": "node -v"
      },
      {
        "runtime": "node -e \"console.log(process.platform)\""
      }
    ]
  }
}
