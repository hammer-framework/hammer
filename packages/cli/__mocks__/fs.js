import { sep } from 'path'

import { isString } from 'lodash'

const fs = {
  ...jest.requireActual('fs'),
}

// This is a custom function that our tests can use during setup to specify
// what the files on the "mock" filesystem should look like when any of the
// `fs` APIs are used.
let mockFiles = {}
/** @param newMockFiles - {[filepath]: contents} */
fs.__setMockFiles = (newMockFiles) => {
  mockFiles = { ...newMockFiles }
}

fs.existsSync = (path) => {
  return isString(mockFiles[path])
}

fs.mkdirSync = () => {}

fs.readFileSync = (path) => {
  // In prisma v4.3.0, prisma format uses a Wasm module. See https://github.com/prisma/prisma/releases/tag/4.3.0.
  // We shouldn't mock this, so we'll use the real fs.readFileSync.
  if (path.includes('prisma_fmt_build_bg.wasm')) {
    return jest.requireActual('fs').readFileSync(path)
  }

  return mockFiles[path]
}

fs.readdirSync = (path) => {
  const mockedFiles = []
  Object.keys(mockFiles).forEach((mockedPath) => {
    // Is a child of the desired path
    if (mockedPath.startsWith(path)) {
      const childPath = mockedPath.substring(path.length + 1)
      // Is the child a direct child of the desired path
      if (!childPath.includes(sep)) {
        mockedFiles.push(childPath)
      }
    }
  })
  return mockedFiles
}

fs.writeFileSync = (path, contents) => {
  mockFiles[path] = contents
}

fs.unlinkSync = (path) => {
  delete mockFiles[path]
}

fs.rmSync = (path) => {
  delete mockFiles[path]
}

// @MARK this works for now but should probably be a little more full-featured
fs.rmdirSync = (path) => {
  delete mockFiles[path]
}

module.exports = fs
