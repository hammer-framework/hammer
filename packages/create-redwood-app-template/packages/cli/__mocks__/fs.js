export const readFileSync = () => {
  console.log('hello')

  return ''
}
