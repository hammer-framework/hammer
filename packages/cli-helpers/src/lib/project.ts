import fs from 'fs'
import path from 'path'

import * as toml from 'toml'

import { resolveFile, findUp } from '@redwoodjs/project-config'
import { getConfigPath } from '@redwoodjs/project-config'

import { colors } from './colors'
import { getPaths } from './paths'

export const getGraphqlPath = () => {
  return resolveFile(path.join(getPaths().api.functions, 'graphql'))
}

export const graphFunctionDoesExist = () => {
  const graphqlPath = getGraphqlPath()
  return graphqlPath && fs.existsSync(graphqlPath)
}

export const isTypeScriptProject = () => {
  const paths = getPaths()
  return (
    fs.existsSync(path.join(paths.web.base, 'tsconfig.json')) ||
    fs.existsSync(path.join(paths.api.base, 'tsconfig.json'))
  )
}

export const getInstalledRedwoodVersion = () => {
  try {
    const packageJson = require('../../package.json')
    return packageJson.version
  } catch (e) {
    console.error(colors.error('Could not find installed redwood version'))
    process.exit(1)
  }
}

/**
 * Updates the project's redwood.toml file to include the specified packages plugin
 *
 * Uses toml parsing to determine if the plugin is already included in the file and
 * only adds it if it is not.
 *
 * Writes the updated config to the file system by appending strings, not stringify-ing the toml.
 */
export const updateTomlConfig = (packageName: string) => {
  const redwoodTomlPath = getConfigPath()

  const configLines = []

  const configContent = fs.readFileSync(redwoodTomlPath, 'utf-8')
  const config = toml.parse(configContent)

  if (!config || !config.experimental || !config.experimental.cli) {
    configLines.push('[experimental.cli]')
  }

  if (!config?.experimental?.cli?.autoInstall) {
    configLines.push('  autoInstall = true')
  }

  if (!config?.experimental?.cli?.plugins) {
    // If plugins array is missing, create it.
    configLines.push('  [[experimental.cli.plugins]]')
  }

  // Check if the package is not already in the plugins array
  if (
    !config?.experimental?.cli?.plugins?.some(
      (plugin: any) => plugin.package === packageName
    )
  ) {
    configLines.push(`    package = "${packageName}"`)
  }

  const newConfig = configContent + configLines.join('\n')

  fs.writeFileSync(redwoodTomlPath, newConfig, 'utf-8')
}

export const updateTomlConfigTask = (packageName: string) => {
  return {
    title: `Updating redwood.toml to configure ${packageName} ...`,
    task: () => {
      updateTomlConfig(packageName)
    },
  }
}

export const addEnvVarTask = (
  name: string,
  value: string,
  comment: string,
  overwrite = false
) => {
  return {
    title: `Adding ${name} var to .env...`,
    task: () => {
      addEnvVar(name, value, comment, overwrite)
    },
  }
}

export const addEnvVar = (
  name: string,
  value: string,
  comment: string,
  overwrite = false
) => {
  const envPath = path.join(getPaths().base, '.env')
  const content = [comment && `# ${comment}`, `${name}=${value}`, ''].flat()
  let envFile = ''

  if (fs.existsSync(envPath)) {
    envFile = fs.readFileSync(envPath).toString()
    const lines = envFile.split('\n')

    // Check if the variable already exists
    const existingIndex = lines.findIndex((line) => {
      const trimmedLine = line.trim()
      return (
        trimmedLine.startsWith(`${name}=`) ||
        trimmedLine.startsWith(`#${name}=`)
      )
    })

    if (existingIndex !== -1) {
      // Variable already exists, check if overwrite is true
      if (overwrite) {
        // Update the existing line with the new value
        const existingComment = [content[0]]
        lines[existingIndex] = `${existingComment}\n${name}=${value}`
        envFile = lines.join('\n')
      }
      // If overwrite is false, do nothing (leave the file unchanged)
    } else {
      // Variable doesn't exist, add it
      envFile += '\n' + content.join('\n')
    }
  } else {
    envFile = content.join('\n')
  }

  return fs.writeFileSync(envPath, envFile)
}

/**
 * This sets the `RWJS_CWD` env var to the redwood project directory. This is typically required for internal
 * redwood packages to work correctly. For example, `@redwoodjs/project-config` uses this when reading config
 * or paths.
 *
 * @param cwd Explicitly set the redwood cwd. If not set, we'll try to determine it automatically. You likely
 * only want to set this based on some specific input, like a CLI flag.
 */
export const setRedwoodCWD = (cwd?: string) => {
  // Get the existing `cwd` from the `RWJS_CWD` env var, if it exists.
  cwd ??= process.env.RWJS_CWD

  if (cwd) {
    // `cwd` was specifically passed in or the `RWJS_CWD` env var. In this case,
    // we don't want to find up for a `redwood.toml` file. The `redwood.toml` should just be in that directory.
    if (!fs.existsSync(path.join(cwd, 'redwood.toml'))) {
      throw new Error(`Couldn't find a "redwood.toml" file in ${cwd}`)
    }
  } else {
    // `cwd` wasn't set. Odds are they're in a Redwood project,
    // but they could be in ./api or ./web, so we have to find up to be sure.
    const redwoodTOMLPath = findUp('redwood.toml', process.cwd())
    if (!redwoodTOMLPath) {
      throw new Error(
        `Couldn't find up a "redwood.toml" file from ${process.cwd()}`
      )
    }
    if (redwoodTOMLPath) {
      cwd = path.dirname(redwoodTOMLPath)
    }
  }

  process.env.RWJS_CWD = cwd
}
