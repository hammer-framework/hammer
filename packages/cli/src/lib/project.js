import fs from 'node:fs'
import path from 'node:path'

import { getPaths } from '.'

export const isTypeScriptProject = () => {
  const paths = getPaths()
  return (
    fs.existsSync(path.join(paths.web.base, 'tsconfig.json')) ||
    fs.existsSync(path.join(paths.api.base, 'tsconfig.json'))
  )
}

export const sides = () => {
  const paths = getPaths()

  let sides = []
  if (fs.existsSync(path.join(paths.web.base, 'package.json'))) {
    sides = [...sides, 'web']
  }
  if (fs.existsSync(path.join(paths.api.base, 'package.json'))) {
    sides = [...sides, 'api']
  }
  return sides
}
