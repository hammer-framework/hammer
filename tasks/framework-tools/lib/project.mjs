/* eslint-env node */

import path from 'node:path'

import execa from 'execa'
import fs from 'fs-extra'
import ora from 'ora'
import { rimraf } from 'rimraf'
import terminalLink from 'terminal-link'

import {
  getFrameworkDependencies,
  getFrameworkPackageJsonPaths,
  getFrameworkPackagesFiles,
  getFrameworkPackagesBins,
  getPackageJsonName,
} from './framework.mjs'

/**
 * Sets binaries as executable and creates symlinks to `node_modules/.bin` if they do not exist.
 */
export function fixProjectBinaries(projectPath) {
  const bins = getFrameworkPackagesBins()

  for (let [binName, binPath] of Object.entries(bins)) {
    // if the binPath doesn't exist, create it.
    const binSymlink = path.join(projectPath, 'node_modules/.bin', binName)
    binPath = path.join(projectPath, 'node_modules', binPath)

    if (!fs.existsSync(binSymlink)) {
      fs.mkdirSync(path.dirname(binSymlink), {
        recursive: true,
      })
      fs.symlinkSync(binPath, binSymlink)
    }

    console.log('chmod +x', terminalLink(binName, binPath))
    fs.chmodSync(binSymlink, '755')
    fs.chmodSync(binPath, '755')
  }
}

/**
 * Add all the `@redwoodjs` packages' dependencies to the root `package.json` in a Redwood Project.
 *
 * @param {string} packageJsonPath - The path to the root `package.json` in a Redwood Project.
 * @param {{ [key: string]: string }?} dependencies - A map of package names to versions.
 *
 * @returns {void}
 */
export function addDependenciesToPackageJson(
  packageJsonPath,
  dependencies = getFrameworkDependencies()
) {
  const packageJsonLink = terminalLink(
    'package.json',
    'file://' + packageJsonPath
  )

  const numberOfDependencies = Object.keys(dependencies).length

  const spinner = ora(
    `Adding ${numberOfDependencies} framework dependencies to ${packageJsonLink}...`
  ).start()

  const packageJson = fs.readJSONSync(packageJsonPath)

  packageJson.dependencies = {
    ...packageJson.dependencies,
    ...dependencies,
  }

  fs.writeJSONSync(packageJsonPath, packageJson, { spaces: 2 })

  spinner.succeed(
    `Added ${numberOfDependencies} framework dependencies to ${packageJsonLink}`
  )
}

export function installProjectPackages(projectPath) {
  const spinner = ora("Running 'yarn install'...")

  spinner.start()

  try {
    execa.commandSync('yarn install', {
      cwd: projectPath,
      shell: true,
    })
    spinner.succeed("Ran 'yarn install'")
  } catch (e) {
    spinner.warn(
      `Error running 'yarn install', check ${terminalLink(
        'yarn-error.log',
        'file://' + path.join(projectPath, 'yarn-error.log')
      )} for more information.`
    )

    console.log('-'.repeat(80))
  }
}

export async function copyFrameworkFilesToProject(
  projectPath,
  packageJsonPaths = getFrameworkPackageJsonPaths()
) {
  // Loop over every package, delete all existing files, copy over the new files,
  // and fix binaries.
  const packagesFiles = await getFrameworkPackagesFiles(packageJsonPaths)

  const packageNamesToPaths = packageJsonPaths.reduce(
    (packageNamesToPaths, packagePath) => {
      packageNamesToPaths[getPackageJsonName(packagePath)] =
        path.dirname(packagePath)
      return packageNamesToPaths
    },
    {}
  )

  for (const [packageName, files] of Object.entries(packagesFiles)) {
    const packageDstPath = path.join(projectPath, 'node_modules', packageName)
    console.log(
      terminalLink(packageName, 'file://' + packageDstPath),
      files.length,
      'files'
    )
    await rimraf(packageDstPath)

    for (const file of files) {
      const src = path.join(packageNamesToPaths[packageName], file)
      const dst = path.join(packageDstPath, file)
      fs.mkdirSync(path.dirname(dst), { recursive: true })
      fs.copyFileSync(src, dst)
    }
  }

  console.log()
  fixProjectBinaries(projectPath)
}
