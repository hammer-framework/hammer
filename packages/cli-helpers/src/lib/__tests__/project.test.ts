import fs from 'fs'

// import toml from 'toml' // Make sure to import 'toml' or an equivalent library

// import { updateTomlConfig, addEnvVar } from '../project' // Replace with the correct path to your module
import { addEnvVar } from '../project' // Replace with the correct path to your module

jest.mock('fs')

jest.mock('@redwoodjs/project-config', () => {
  return {
    getPaths: () => {
      return {
        generated: {
          base: '.redwood',
        },
        base: '',
      }
    },
    getConfig: jest.fn(),
  }
})

describe('addEnvVar adds environment variables as part of a setup task', () => {
  let envFileContent = ''

  describe('addEnvVar adds environment variables as part of a setup task', () => {
    beforeEach(() => {
      jest.spyOn(fs, 'existsSync').mockImplementation(() => {
        return true
      })

      jest.spyOn(fs, 'readFileSync').mockImplementation(() => {
        return envFileContent
      })

      jest.spyOn(fs, 'writeFileSync').mockImplementation((envPath, envFile) => {
        expect(envPath).toContain('.env')
        return envFile
      })
    })

    afterEach(() => {
      // Restore the original implementations of fs.existsSync and fs.readFileSync
      jest.restoreAllMocks()
      envFileContent = ''
    })

    it('should add a new environment variable when it does not exist (overwrite: false)', () => {
      envFileContent = 'EXISTING_VAR=value\n# CommentedVar=123\n'
      const file = addEnvVar(
        'NEW_VAR',
        'new_value',
        'New Variable Comment',
        false
      )
      expect(file).toMatchSnapshot()
    })

    it('should updates an existing environment variable when it exists and overwrite chosen', () => {
      envFileContent = 'EXISTING_VAR=value\n# CommentedVar=123\n'
      const file = addEnvVar(
        'EXISTING_VAR',
        'new_value',
        'Updated existing variable Comment',
        true
      )
      expect(file).toMatchSnapshot()
    })

    it('should not update existing environment variable if exists and overwrite is default', () => {
      envFileContent = 'EXISTING_VAR=value\n# CommentedVar=123\n'
      const file = addEnvVar(
        'EXISTING_VAR',
        'new_value',
        'Updated existing variable Comment'
      )
      expect(file).toMatchSnapshot()
    })

    it('should not update existing environment variable if exists and overwrite is false', () => {
      envFileContent = 'EXISTING_VAR=value\n# CommentedVar=123\n'
      const file = addEnvVar(
        'EXISTING_VAR',
        'new_value',
        'Updated existing variable Comment',
        false
      )
      expect(file).toMatchSnapshot()
    })
  })
})

// describe('updateTomlConfig', () => {
//   const redwoodTomlPath = 'fake-path-to-redwood-toml'
//   const packageName = 'test-package-name'

//   beforeEach(() => {
//     // Reset the mocked functions and data
//     ;(fs.readFileSync as jest.Mock).mockClear()
//     ;(fs.writeFileSync as jest.Mock).mockClear()
//     ;(toml.parse as jest.Mock).mockClear()
//   })

//   test('it should update the configuration with a new package', () => {
//     const existingConfig = {
//       experimental: {
//         cli: {
//           plugins: [{ package: 'existing-package' }],
//         },
//       },
//     }

//     const expectedNewConfig = {
//       experimental: {
//         cli: {
//           plugins: [{ package: 'existing-package' }, { package: packageName }],
//         },
//       },
//     }

//     ;(fs.readFileSync as jest.Mock).mockReturnValue(
//       JSON.stringify(existingConfig)
//     )
//     ;(toml.parse as jest.Mock).mockReturnValue(existingConfig)

//     updateTomlConfig(packageName)

//     expect(fs.readFileSync).toHaveBeenCalledWith(redwoodTomlPath, 'utf-8')
//     expect(fs.writeFileSync).toHaveBeenCalledWith(
//       redwoodTomlPath,
//       toml(expectedNewConfig),
//       'utf-8'
//     )
//   })

//   test('it should add missing configuration sections and the package', () => {
//     // Simulate a missing 'experimental.cli' section in the configuration
//     const existingConfig = {}

//     const expectedNewConfig = {
//       experimental: {
//         cli: {
//           autoInstall: true,
//           plugins: [{ package: packageName }],
//         },
//       },
//     }

//     fs.readFileSync.mockReturnValue('')
//     toml.parse.mockReturnValue(existingConfig)

//     updateTomlConfig(packageName)

//     expect(fs.readFileSync).toHaveBeenCalledWith(redwoodTomlPath, 'utf-8')
//     expect(fs.writeFileSync).toHaveBeenCalledWith(
//       redwoodTomlPath,
//       toml.dump(expectedNewConfig),
//       'utf-8'
//     )
//   })

//   // Add more test cases for different scenarios (e.g., autoInstall already set, etc.)
// })
