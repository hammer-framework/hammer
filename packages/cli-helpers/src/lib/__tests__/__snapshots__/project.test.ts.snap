// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addEnvVar adds environment variables as part of a setup task addEnvVar adds environment variables as part of a setup task should add a new environment variable when it does not exist (overwrite: false) 1`] = `
"EXISTING_VAR=value
# CommentedVar=123

# New Variable Comment
NEW_VAR=new_value
"
`;

exports[`addEnvVar adds environment variables as part of a setup task addEnvVar adds environment variables as part of a setup task should not update existing environment variable if exists and overwrite is default 1`] = `
"EXISTING_VAR=value
# CommentedVar=123
"
`;

exports[`addEnvVar adds environment variables as part of a setup task addEnvVar adds environment variables as part of a setup task should not update existing environment variable if exists and overwrite is false 1`] = `
"EXISTING_VAR=value
# CommentedVar=123
"
`;

exports[`addEnvVar adds environment variables as part of a setup task addEnvVar adds environment variables as part of a setup task should updates an existing environment variable when it exists and overwrite chosen 1`] = `
"# Updated existing variable Comment
EXISTING_VAR=new_value
# CommentedVar=123
"
`;
